src/main/java/cs/f10/t1/nursetraverse/logic/parser/PatientBookParser.java

```diff
@@ -55,111 +58,86 @@
      * @return the command based on the user input
      * @throws ParseException if the user input does not conform the expected format
      */
-    public Command parseCommand(String userInput) throws ParseException {
+    public Command parseCommand(String userInput, ModelState modelState) throws ParseException {
         final Matcher matcher = BASIC_COMMAND_FORMAT.matcher(userInput.trim());
         if (!matcher.matches()) {
             throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, HelpCommand.MESSAGE_USAGE));
         }
 
         final String commandWord = matcher.group("commandWord");
         final String arguments = matcher.group("arguments");
-        Command command;
-        switch (commandWord) {
+        Command command = parseCommandByCommandWord(commandWord, arguments, modelState);
+        command.setCommandText(userInput);
+        return command;
+    }
 
+    /**
+     * Helper method to create command by commandWord.
+     */
+    private Command parseCommandByCommandWord(final String commandWord, final String arguments, ModelState modelState)
+            throws ParseException {
+        switch (commandWord) {
         case AddCommand.COMMAND_WORD:
-            command = new AddCommandParser().parse(arguments);
-            break;
-
+            return new AddCommandParser().parse(arguments);
         case EditCommand.COMMAND_WORD:
-            command = new EditCommandParser().parse(arguments);
-            break;
-
+            return new EditCommandParser().parse(arguments);
         case DeleteCommand.COMMAND_WORD:
-            command = new DeleteCommandParser().parse(arguments);
-            break;
-
+            return new DeleteCommandParser().parse(arguments);
         case ClearCommand.COMMAND_WORD:
-            command = new ClearCommand();
-            break;
-
+            return new ClearCommand();
         case FindCommand.COMMAND_WORD:
-            command = new FindCommandParser().parse(arguments);
-            break;
-
+            return new FindCommandParser().parse(arguments);
         case ListCommand.COMMAND_WORD:
-            command = new ListCommand();
-            break;
-
+            return new ListCommand();
         case ExitCommand.COMMAND_WORD:
-            command = new ExitCommand();
-            break;
-
+            return new ExitCommand();
         case HelpCommand.COMMAND_WORD:
-            command = new HelpCommand();
-            break;
-
+            return new HelpCommand();
         case ExportCommand.COMMAND_WORD:
-            command = new ExportCommandParser().parse(arguments);
-            break;
-
+            return new ExportCommandParser().parse(arguments);
         case ImportReplaceCommand.COMMAND_WORD:
-            command = new ImportReplaceCommandParser().parse(arguments);
-            break;
-
+            if (modelState.equals(ModelState.VISIT_ONGOING)) {
+                throw new ParseException(MESSAGE_REQUIRE_STATE_NORMAL);
```

Comment:
ParseException has more to do with command syntax, maybe something like `StateException` will be more accurate.